analysis/syntactic/preconditions/precond_check.go

```
@@ -0,0 +1,287 @@
+		return false
+	}
+	for callee := range callees {
+		if fn.Exists(spec.Call, func(c config.CodeIdentifier) bool {
+			return c.MatchPackageAndMethod(instr.Parent(), callee)
+		}) {
+			return true
+		}
+	}
+	return false
+}
+
+// blockCond optionally returns an atomic proposition for a block given as argument. It only returns
+// Some string-represented proposition if and only if the last statement of the block is an If.
+func blockCond(block *ssa.BasicBlock) fn.Optional[string] {
+	if len(block.Instrs) <= 0 {
+		return fn.None[string]()
+	}
+	ifCond, lastInstrIsIf := block.Instrs[len(block.Instrs)-1].(*ssa.If)
+	if !lastInstrIsIf {
+		return fn.None[string]()
+	}
+	return fn.Some(stringCond(ifCond.Cond))
+}
+
+// stringCond returns a string representation of the conditional. It attempts to convert the value to an expression
+// by recursively looking at its operands. When some operand cannot be converted to a string representation of an
+// expression, then it is "<!nae!>" (not an expression).
+// Only function call that are not "invokes" are converted, and their arguments are ommitted, e.g. a call to f with
+// some arguments is always printed "f(...)". This means the condition-check cannot currently match on specific
+// arguments.
+//
+// TODO: this will be replaced by a proper representation for expressions that can be compared and manipulated.
+func stringCond(v ssa.Value) string {
+	switch value := v.(type) {
+	case *ssa.Call:
+		if value.Common().IsInvoke() {
+			return value.Common().Method.Name() + "(...)"
+		}
+		return value.Common().Value.Name() + "(...)"
+	case *ssa.Extract:
+		return fmt.Sprintf("%s#%d", stringCond(value.Tuple), value.Index)
+	case *ssa.BinOp:
+		return fmt.Sprintf("%s %s %s", stringCond(value.X), value.Op, stringCond(value.Y))
+	case *ssa.UnOp:
+		return fmt.Sprintf("%s %s", value.Op, stringCond(value.X))
+	case *ssa.Const:
+		return value.String()
+	default:
+		return "<!nae!>"
```

Would a panic or error value be more appropriate here to avoid silent failures?

---

It's expected that those will occur. For example, if I need to check that A and B are true when I call f(), it's fine if `A & B & <!nae!>` holds. The analysis also doesn't make guarantees on side-effects, it's just about what has been executed on the path and the result of some function calls, so it's ok that there are uninterpreted literals in the path formulas.